Some more drive-by WebGL fixes and cleanups

nox opened the pull request:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/21523)
<!-- Reviewable:end -->

nox committed:
Fix the error for invalid arrays passed to gl.vertexAttrib*v()

nox committed:
Allow binding the null buffer in vertexAttribPointer if offset is 0

nox committed:
Clean up some WebGL expectations

nox committed:
Remove WebGLRenderingContext.STENCIL_INDEX

nox committed:
Disable quickCheckAPI-S_V.html

It generates random input which even trips the invalid surrogate panic.

nox committed:
Disable rapid-resizing.html

The absence of support for preserveDrawingBuffer makes it timeout.

nox committed:
Always emit INVALID_OPERATION on null element buffers in drawElements

nox committed:
Fix gl.isBuffer for buffers that are marked for deletion but still attached

highfive commented:
Heads up! This PR modifies the following files:
 * @asajeffrey: components/script/dom/webglvertexarrayobjectoes.rs, components/script/dom/webglrenderingcontext.rs, components/script/dom/webidls/WebGLRenderingContext.webidl
 * @jgraham: tests/wpt/webgl/meta/conformance/rendering/preservedrawingbuffer-leak.html.ini, tests/wpt/webgl/meta/conformance/renderbuffers/framebuffer-state-restoration.html.ini, tests/wpt/webgl/meta/conformance/canvas/to-data-url-test.html.ini, tests/wpt/webgl/meta/conformance/attribs/gl-vertex-attrib.html.ini, tests/wpt/webgl/meta/conformance/more/conformance/methods.html.ini and 43 more
 * @kichjang: components/script/dom/webglvertexarrayobjectoes.rs, components/script/dom/webglrenderingcontext.rs, components/script/dom/webidls/WebGLRenderingContext.webidl

nox commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit 26f90daabcff0ec213594b93920ba8cd60fd8f1d with merge 0d45913d584bc7302f858d2a8bb97b10db8caa79...

bors-servo commented:
:broken_heart: Test failed - [mac-rel-css1](http://build.servo.org/builders/mac-rel-css1/builds/5115)

nox committed:
Disable many texture-related tests for things we don't support yet

nox commented:
Disabled many more tests, notably everything related to texture sources that we don't support at all.

@bors-servo try

bors-servo commented:
:hourglass: Trying commit eb204573e0884d12fed76e4f6b6b6a8dd689217b with merge 59f9b1d5239f8e3ae841413fb7f5b5a637593134...

bors-servo commented:
:broken_heart: Test failed - [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/9427)

nox commented:
Only unrelated failures again.

nox committed:
Fix Drop assertion for WebGLBuffer

A buffer that is dropped may not have been marked for deletion but it
must not be attached to any VAO.

nox committed:
Delete our Servo-specific WebGL tests

Those test things that are already tested by Khronos tests.

nox committed:
Support unions of objects in overloads

Part of #20513, implementing the parts useful for WebGL.

nox committed:
Make setZeroTimeout use window.setTimeout

nox commented on tests/wpt/webgl/meta/conformance/textures/canvas/__dir__.ini:
```diff
@@ -0,0 +1 @@
+disabled: canvas textures are not implemented yet
```
I was actually completely wrong about this specific folder of tests.

nox reviewed (commented):
I still need to add a few issue links but it's gym time.

nox committed:
Remove obsolete patch

nox committed:
Merge unit.patch and unit2.patch

nox committed:
Make use of window.onerror in Khronos tests

avadacatavra commented on tests/wpt/webgl/meta/conformance/textures/misc/tex-video-using-tex-unit-non-zero.html.ini:
```diff
@@ -1,8 +1,3 @@
-
-  [Overall test]
-    expected: NOTRUN
+  disabled: video textures are not supported yet
```
file and link to issue

avadacatavra commented on tests/wpt/webgl/meta/conformance/textures/misc/texture-corner-case-videos.html.ini:
```diff
@@ -1,5 +1,3 @@
-  expected: TIMEOUT
-  [Overall test]
-    expected: NOTRUN
+  disabled: video textures are not supported yet
```
link to issue

avadacatavra commented on tests/wpt/webgl/meta/conformance/textures/misc/texture-npot-video.html.ini:
```diff
@@ -1,5 +1,3 @@
-  expected: TIMEOUT
-  [Overall test]
-    expected: NOTRUN
+  disabled: video textures are not supported yet
```
link to issue

avadacatavra commented on tests/wpt/webgl/meta/conformance/textures/svg_image/__dir__.ini:
```diff
@@ -0,0 +1 @@
+disabled: SVG support is not implemented yet
```
iirc we don't have plans to implement this right now, so if that's the case I don't see the need to file an issue for it

avadacatavra commented on tests/wpt/webgl/meta/conformance/textures/video/__dir__.ini:
```diff
@@ -0,0 +1 @@
+disabled: video textures are not implemented yet
```
link to issue

avadacatavra commented on tests/wpt/webgl/meta/conformance/textures/webgl_canvas/__dir__.ini:
```diff
@@ -0,0 +1 @@
+disabled: canvas textures are not implemented yet
```
file/link to issue

avadacatavra commented on tests/wpt/webgl/meta/conformance/canvas/draw-static-webgl-to-multiple-canvas-test.html.ini:
```diff
@@ -1,5 +1,3 @@
-  expected: ERROR
-  [Overall test]
-    expected: NOTRUN
+  disabled: drawing webgl canvas in 2d canvas is not supported yet
```
link to issue

avadacatavra commented on tests/wpt/webgl/meta/conformance/canvas/draw-webgl-to-canvas-test.html.ini:
```diff
@@ -1,5 +1,3 @@
-  expected: ERROR
-  [Overall test]
-    expected: NOTRUN
+  disabled: drawing webgl canvas in 2d canvas is not supported yet
```
link to issue

nox commented:
@jdm You need to sign on two things: disabling the quickcheck test and deleting the Servo-specific WebGL tests. @avadacatavra reviewed everything else.

bors-servo commented:
:hourglass: Testing commit 63076413817055dd32a66939a9c533baaa4af9bb with merge 4fec3335b9604b135d2d0bcb008689f7182b273e...

jdm commented on tests/wpt/webgl/tests/js/js-test-pre.js:
```diff
@@ -143,8 +143,10 @@ function notifyFinishedToHarness() {
+  var oldOnError = window.onerror;
   var t = async_test("Overall test");
   window.notifyFinishedToHarness = t.step_func_done(oldNotify);
+  window.onerror = t.step_func_done(oldNotify);
```
This doesn't use oldOnError.

nox replied:
I used `addEventListener` instead.

jdm commented:
@bors-servo r-

bors-servo commented:
:broken_heart: Test failed - [mac-rel-css2](http://build.servo.org/builders/mac-rel-css2/builds/5129)

bors-servo commented:
:hourglass: Testing commit 900a19058e5f282fdb914ac16801d11f4a2c0159 with merge 8af3bf28d32a4178ec1d0c6746d313fad5e6d46a...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/9600)

jdm commented:
@bors-servo retry
* infra problems

bors-servo commented:
:broken_heart: Test failed - [linux-dev](http://build.servo.org/builders/linux-dev/builds/12665)

jdm commented:
@bors-servo retry
* busted build machine

bors-servo commented:
:zap: Previous build results for [android-x86](http://build.servo.org/builders/android-x86/builds/479), [mac-rel-css2](http://build.servo.org/builders/mac-rel-css2/builds/5136), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/9447), [mac-rel-wpt3](http://build.servo.org/builders/mac-rel-wpt3/builds/5209) are reusable. Rebuilding only [android](http://build.servo.org/builders/android/builds/12368), [arm32](http://build.servo.org/builders/arm32/builds/12562), [arm64](http://build.servo.org/builders/arm64/builds/12604), [linux-dev](http://build.servo.org/builders/linux-dev/builds/12665), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/9602), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/9605), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/12693), [mac-rel-css1](http://build.servo.org/builders/mac-rel-css1/builds/5138), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/9405), [mac-rel-wpt4](http://build.servo.org/builders/mac-rel-wpt4/builds/5150), [windows-msvc-dev](http://build.servo.org/builders/windows-msvc-dev/builds/8068)...

bors-servo commented:
:broken_heart: Test failed - [mac-rel-css1](http://build.servo.org/builders/mac-rel-css1/builds/5138)

jdm commented:
@bors-servo retry

bors-servo commented:
:zap: Previous build results for [android](http://build.servo.org/builders/android/builds/12368), [android-x86](http://build.servo.org/builders/android-x86/builds/479), [arm32](http://build.servo.org/builders/arm32/builds/12562), [arm64](http://build.servo.org/builders/arm64/builds/12604), [linux-dev](http://build.servo.org/builders/linux-dev/builds/12666), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/9602), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/9605), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/12693), [mac-rel-css2](http://build.servo.org/builders/mac-rel-css2/builds/5136), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/9447), [mac-rel-wpt3](http://build.servo.org/builders/mac-rel-wpt3/builds/5209), [windows-msvc-dev](http://build.servo.org/builders/windows-msvc-dev/builds/8068) are reusable. Rebuilding only [mac-rel-css1](http://build.servo.org/builders/mac-rel-css1/builds/5138), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/9405), [mac-rel-wpt4](http://build.servo.org/builders/mac-rel-wpt4/builds/5150)...

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/12368), [android-x86](http://build.servo.org/builders/android-x86/builds/479), [arm32](http://build.servo.org/builders/arm32/builds/12562), [arm64](http://build.servo.org/builders/arm64/builds/12604), [linux-dev](http://build.servo.org/builders/linux-dev/builds/12666), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/9602), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/9605), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/12693), [mac-rel-css1](http://build.servo.org/builders/mac-rel-css1/builds/5140), [mac-rel-css2](http://build.servo.org/builders/mac-rel-css2/builds/5136), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/9405), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/9447), [mac-rel-wpt3](http://build.servo.org/builders/mac-rel-wpt3/builds/5209), [mac-rel-wpt4](http://build.servo.org/builders/mac-rel-wpt4/builds/5150), [windows-msvc-dev](http://build.servo.org/builders/windows-msvc-dev/builds/8068)
Approved by: avadacatavra,jdm
Pushing 6f81ed49f6a7281341d079766de0601286c4802c to master...
